File: imperative/input/instruction.py
from input.token import Token

class Instruction:
    instruction_types = {
        "invalid": -1,
        "binary_operator": 0,
        "unary_operator": 1,
        "assignment": 2
    }

    def __init__(self, type: int, dest: Token, operand1: Token=None, operator: Token=None, operand2: Token=None):
        self.type: int = type
        self.dest: Token = dest
        self.operand1: Token = operand1
        self.operator: Token = operator
        self.operand2: Token = operand2

    def __str__(self):
        match self.type:
            case 0:
                return f"{self.dest.value} = {self.operand1.value} {self.operator.value} {self.operand2.value}"
            case 1:
                return f"{self.dest.value} = {self.operator.value}{self.operand2.value}"
            case 2:
                return f"{self.dest.value} = {self.operand1.value}"
            
    def get_variables(self) -> list[Token]:
        """
        Retrieves all variable tokens involved in this instruction.
        
        :return: A list of variable tokens.
        :rtype: list[Token]
        """

        variable_type = 1 # Token type for variables
        variables: list[Token] = []

        variables.append(self.dest)
        variables.append(self.operand1) if self.operand1 and self.operand1.type == variable_type else None
        variables.append(self.operand2) if self.operand2 and self.operand2.type == variable_type else None

        return variables

File: haskell/input/Instruction.hs
module Instruction (Instruction, Instructions, getVariables) where

import Token

type Dest = Token
type Operand = Token
type Operator = Token

data InstructionType 
    = Invalid
    | BinaryOperator
    | UnaryOperator
    | Assignment
    deriving (Show, Eq)

data Instruction 
    = BinaryIns InstructionType Dest Operand Operator Operand
    | UnaryIns InstructionType Dest Operator Operand
    | AssignmentIns InstructionType Dest Operand
    deriving (Eq)

instance Show Instruction where
    show (BinaryIns _ dest op1 operator op2) = "Binary Instruction: " ++ show dest ++ " = " ++ show op1 ++ " " ++ show operator ++ " " ++ show op2
    show (UnaryIns _ dest operator op) = "Unary Instruction: " ++ show dest ++ " = " ++ show operator ++ " " ++ show op
    show (AssignmentIns _ dest op) = "Assignment Instruction: " ++ show dest ++ " = " ++ show op

instance Show [Instruction] where
    show instructions = "Instructions: \n" ++ map (\x -> show x ++ "\n") instructions

data LiveVariable = Live String

instance Show LiveVariable where
    show (Live var) = var ++ ","

instance Show [LiveVariable] where
    show liveVars = "Live: " ++ map show liveVars

-- Doesn't have to be a pair
data Instructions
    = Instruction ([Instruction],[LiveVariable])

instance Show Instructions where
    show (Instruction (ins,live)) = show ins ++ show live

-- Public: Retrieves all variable tokens stored in the given instruction
getVariables :: Instruction -> [Token]
getVariables (BinaryIns _ dest op1 _ op2) = filter isVariable [dest, op1, op2]
getVariables (UnaryIns _ dest _ op) = filter isVariable [dest, op]
getVariables (AssignmentIns _ dest op) = filter isVariable [dest, op]

-- Private: Helper that checks if Token is a variable
isVariable :: Token -> Bool
isVariable (Token _ Variable) = True
isVariable _ = False